### ISSUE TITLE
feat: enable the substrait compiler to compile substrait protobuf extensions (such as custom relations)

### ISSUE THREAD

**Issue by drin** [ibis-project/ibis-substrait#498]
## Request
General feature request to enable `ibis_substrait.compiler.core.SubstraitCompiler` (or some similar interface) to compile substrait plans with extensions.

I have minimal opinion on approaches, but I am planning on using extensions and I thought it would be nice for others who use ibis to be able to easily leverage their own extensions (instead of having to rewrite the compiler for the core specification).

## Currently
Usage seems to compile an ibis expression using the core substrait specification:
```python
from ibis_substrait.compiler.core import SubstraitCompiler


compiler = SubstraitCompiler()

# where `expr` is some ibis expression
proto = compiler.compile(expr)
```

## Future
For applications and systems that use substrait extensions, it would be nice to be able to accept some extension-specific compiler(s) to compile as appropriate for the extension specification:
```python
from ibis_substrait.compiler.core import SubstraitCompiler


class SomeExtensionCompiler:
    """
    A compiler that looks like `SubstraitCompiler` but knows how to compile some substrait extension(s).
    """
    pass


compiler = SubstraitCompiler()

# We might want to just add another compiler that does its own pass first
# Or perhaps we would also want to specify if it does a preceding or proceeding pass?
compiler.RegisterCompiler(SomeExtensionCompiler())

# where `expr` is some ibis expression
proto = compiler.compile(expr)
```

Edit: updated title to differentiate this feature from function extensions given this comment:
> I will note that at this point, our extension support is limited to the yaml extensions that define scalar and aggregate functions.

**Comment by drin** [comment]
I'm not sure if there's a desired format, so let me know if I should change anything for convenience.

Also, for reference, I am going to try and prototype something using these repositories:
* [mohair](https://github.com/drin/mohair) -- prototype of a query processor
* [mohair-protocol](https://github.com/drin/mohair-protocol) -- my own substrait extension(s)

Ideally, using separate repositories will also serve as an example of how others might leverage this feature.

Also, using [buf](https://buf.build/), I am able to compile just my substrait extension without compiling the core substrait specification, in which case it may (should) be possible for the generated code to rely on the protobuf wrappers included in the ibis-substrait package.

**Comment by drin** [comment]
decided to make a new repo just for prototyping for this feature: [mohair-extension](https://github.com/drin/mohair-extension/).

this repo is the same as mohair, but will try to avoid:
* pulling in the substrait specification
* housing the compiled substrait python wrappers

and will try to lean on the ibis-substrait more heavily. This will allow me to keep prototyping whatever I need in the other one repo.

**Comment by gforsyth** [comment]
Hey @drin, thanks for opening this!

So there are a few things we need to think about here, I think.  There are a few ways we can extend the existing compiler (or add on to it, register, etc.)

Currently, the `translate.py` file performs all the mappings between ibis ops and the corresponding substrait.
One thing that is currently possible is to import the dispatcher used in `translate.py` and then register additional rules for handling the ops in question.

That only works if the op isn't already being dispatched on -- I'm not sure what the behavior is if you try to overwrite an existing rule (it probably works, but I need to confirm).

We can subclass `SubstraitCompiler`, but it currently imports the `translate` dispatcher and calls that, so I'm not sure if that provides much benefit (with the current setup).

I'm planning on adding in proper support for the bundled Substrait extensions, via some kind of `register` api, where you hand it a YAML file and it parses out the function names, options, and allowed datatypes, so that plans can refer to the correct extension spec file.  

Taking a quick look at the mohair-extension -- I think it would be useful to try adding support for that by adding the appropriate `@translate` rules.

> Also, using [buf](https://buf.build/), I am able to compile just my substrait extension without compiling the core substrait specification, in which case it may (should) be possible for the generated code to rely on the protobuf wrappers included in the ibis-substrait package.

I'm interested to see how this goes -- the thing that may trip us up is that we namespace the substrait protos to avoid collisions because Python / C++ proto stuff interacts in unpleasant ways (they share a descriptor pool)

**Comment by drin** [comment]
> One thing that is currently possible is to import the dispatcher used in translate.py and then register additional rules for handling the ops in question.
> ...
> I think it would be useful to try adding support for that by adding the appropriate @translate rules.

I think this is a good idea and a good start. Essentially, for disjoint ops we can just add new dispatch rules.

I can try playing with the idea of a "RegisteredPasses" class that accepts `SubstraitCompiler`-like classes that uses function dispatch. The fallback can be to delegate to a `SubstraitCompiler` instance. I'll try to see how maintainable it may be and see how it affects performance.

> I'm interested to see how this goes -- the thing that may trip us up is that we namespace the substrait protos...

I am hoping that trying this approach will help us find where the real pain points are so that we can help figure out potential "best practices" for the future (for some definition of "best").

**Comment by drin** [comment]
I have at least made progress on some infrastructure for me to create and parse substrait, but I haven't yet added translation functions that ibis-substrait will interact with. I will try to do this a bit at the end of this week, and see how it looks

**Comment by gforsyth** [comment]
definitely keep me posted.  Also you can take a look at #366 for a version of extension registration (still in progress)

**Comment by drin** [comment]
At the moment, I believe the substrait compiler will always turn a table into an existing type of `ReadRel` (e.g. `NamedTable`), and specifically "a table" some (derived) instance of an ibis table.

is there a way to use my own "table class" so that when I use the ibis substrait compiler it will look that type up in a substrait extension I specify? or is it preferred for me to redefine the appropriate translate function, check if the type is appropriate for my specific translation, and then delegate back to the original translate function otherwise (e.g. [compiler.translate.unbound_table](https://github.com/ibis-project/ibis-substrait/blob/main/ibis_substrait/compiler/translate.py#L711))

**Comment by drin** [comment]
(I'm at a point where I can translate an ibis query into substrait and then execute it with Acero, so now I need to figure out how to do it with things that are not in core substrait)

**Comment by gforsyth** [comment]
I think that a reasonably resilient approach would be to subclass the `UnboundTable` op, then register a corresponding `translate` function to handle that particular table class.

I will note that at this point, our extension support is limited to the yaml extensions that define scalar and aggregate functions.

**Comment by drin** [comment]
I assume this would be where UnboundTable is defined: [relations.py:UnboundTable](https://github.com/ibis-project/ibis/blob/master/ibis/expr/operations/relations.py#L52-L54)?

Is there anything that should be defined, or can it be an empty derived class just for the purposes of type resolution, for example:
```
class SkyTable(UnboundTable): pass
```

I was thinking that it seemed like schema may need to be defined, but I now realize that it maybe just provides an empty Schema type when accessed (assuming that `coerced_to` by default gets `None` if no arg is provided)?

**Comment by gforsyth** [comment]
There will need to be something for `schema` with an `UnboundTable`, but you could use `NamedTable` instead, which doesn't have any required attributes

**Comment by drin** [comment]
> I will note that at this point, our extension support is limited to the yaml extensions that define scalar and aggregate functions.

I got that impression after looking at #510 . I will eventually want to use those types of extensions, but I suspect that what I may want to pass to custom functions would be atypical types such as a SkyRel or a Tensor. Are there plans to support other aspects of extensions in the near-ish future?

**Comment by gforsyth** [comment]
It's very much a needs-driven affair at the moment and currently very little of our work is touching custom protobuf extensions.

**Comment by drin** [comment]
> but you could use NamedTable instead

I can't find NamedTable other than the protobuf wrapper ([search ibis-substrait](https://github.com/search?q=repo%3Aibis-project%2Fibis-substrait%20NamedTable&type=code), [search ibis](https://github.com/search?q=repo%3Aibis-project%2Fibis+NamedTable&type=code)). I see that [PhysicalTable](https://github.com/ibis-project/ibis/blob/master/ibis/expr/operations/relations.py#L47-L48) doesn't have any attributes, but I assume that's not the preferred class to derive from?

**Comment by gforsyth** [comment]
> I can't find NamedTable other than the protobuf wrapper ([search ibis-substrait](https://github.com/search?q=repo%3Aibis-project%2Fibis-substrait%20NamedTable&type=code), [search ibis](https://github.com/search?q=repo%3Aibis-project%2Fibis+NamedTable&type=code)). I see that [PhysicalTable](https://github.com/ibis-project/ibis/blob/master/ibis/expr/operations/relations.py?rgh-link-date=2023-03-14T21%3A35%3A23Z#L47-L48) doesn't have any attributes, but I assume that's not the preferred class to derive from?

Sorry, I did mean `PhysicalTable` (it's poorly named) -- which itself derives from `Named` (hence the brain fart).

`PhysicalTable` is a reasonable choice to subclass.

**Comment by drin** [comment]
haha, I thought that was possible, but I wasn't sure. Thanks!

**Comment by drin** [comment]
hm, so I have tried a variety of approaches but I can't define a custom class that derives from PhysicalTable without getting one of the following errors:
1. got an unexpected keyword argument 'domain'
1. Can't instantiate abstract class SkyTable with abstract methods name, schema
1. `Attribute 'skyrel' cannot be assigned to immutable instance of type <class 'mohair_extension.extension.SkyTable'>`

As far as I can tell, when deriving from `PhysicalTable`:
1. the `__init__` method on the class is not as expected (I can't specify an argument by name)
2. I must define `name()` (required by [Named][src-Named]) and `schema()` (required by [TableNode][src-TableNode]) methods
3. I cannot modify instance attributes (because `TableNode` is a [Node][src-Node] is `Concrete` is `Immutable`)

I'm not sure how to create a class that derives from `PhysicalTable` but holds the right state to be used by the `translate` dispatcher. I suspect I need to leverage something about the `Annotable` class (and I think it needs to be something that [AnnotableMeta][src-AnnotableMeta], specifically, catches), but I was hoping to get a bit more guidance.


[src-Named]:      https://github.com/ibis-project/ibis/blob/343fb375d11e612807061818c612d387903ca469/ibis/expr/operations/core.py#L42-L54
[src-TableNode]: https://github.com/ibis-project/ibis/blob/master/ibis/expr/operations/relations.py#L37-L45
[src-Node]:          https://github.com/ibis-project/ibis/blob/master/ibis/expr/operations/core.py#L17-L18
[src-AnnotableMeta]: https://github.com/ibis-project/ibis/blob/master/ibis/common/grounds.py#L82-L88

**Comment by drin** [comment]
The current version I have (deriving from `PhysicalTable`) is:
    https://github.com/drin/mohair-extension/blob/mainline/mohair_extension/extension.py#L60-L73

which is executed here:
    https://github.com/drin/mohair-extension/blob/mainline/toolbox/compile_mohair#L154-L160

and used here:
    https://github.com/drin/mohair-extension/blob/mainline/mohair_extension/extension.py#L79-L96

**Comment by gforsyth** [comment]
All of the table node things are immutable by design, so that makes this tricky without resorting to (mild) hacks.
But I'm a fan of mild hacks.  I think you can do something like:

```python
import ibis.expr.rules as rlz

class SkyTable(relations.PhysicalTable):
    skyrel = rlz.optional(rlz.any)
    
    def __init__(self, **kwargs):
        super().__init__(**kwargs)
    
    def SetPartition(self, sky_partition):
        object.__setattr__(self, "skyrel", sky_partition)
        return self
        
    def name(self):
        return f"{self.skyrel.domain.key}/{self.skyrel.meta.key}"
        
    def schema(self):
        return self.skyrel.meta.schema
```

**Comment by drin** [comment]
Wooh! I had to make a bunch of random attempts to figure out how to make it 'subscriptable' (I think just use the `to_expr` function).

I am now at this error, which I think is a good sign (and probably won't be too hard to figure out):
```
TypeError: Parameter to MergeFrom() must be instance of same class: expected google.protobuf.any_pb2.Any got mohair.algebra_pb2.SkyRel. for field ExtensionLeafRel.detail
```

Thanks!

**Comment by drin** [comment]
success:
```python
relations {
  root {
    input {
      project {
        common {
          emit {
            output_mapping: 3
            output_mapping: 4
          }
        }
        input {
          extension_leaf {
            common {
              direct {
              }
            }
            detail {
              [type.googleapis.com/mohair.SkyRel] {
                domain: "public"
                partition: "E-GEOD-100618"
                slices: 0
                slices: 1
                execstats {
                }
              }
            }
          }
        }
...
```

**Comment by drin** [comment]
Okay, I updated the `mohair_extension` repo and I updated the README to (at least minimally) describe what I've prototyped: https://github.com/drin/mohair-extension/#custom-ibis-table-classes

I'll update it a bit more eventually, but I just wanted to see if there's anything you'd like to see in that repo before I close this issue. I think a function extension could be a separate issue, and I am not sure if there's a place to put an example in the ibis-substrait repo.

I also have examples for extending `UnboundTable` and `PhysicalTable`. In general, since ibis expressions are eager, it doesn't seem possible to construct a query without a schema, so I found deriving from `PhysicalTable` to not be practically useful (though it's definitely possible).
